src/codegen/MHTranslator.cpp

```diff
@@ -1036,24 +1036,148 @@ void MHTranslator::transQuery(code::FunctionDecl* fun, std::shared_ptr<ir::Query
-            BOOL_TYPE : mapIRTypeToCodeType(qr->getVar()->getTyp())) })),
-      initvalue);
-  std::vector<code::Expr*> args;
-  args.push_back(transExpr(qr->getVar()));
-  args.push_back(new code::IntegerLiteral(1));
-  fun->addStmt(
-      code::CallExpr::createMethodCall(answervarname, HISTOGRAM_ADD_METHOD_NAME,
-                                       args));
-  // add print this result in print()
-  // :::=> answer_id.print("query expr");
-  corePrint->addStmt(
-      code::CallExpr::createMethodCall(answervarname,
-                                       HISTOGRAM_PRINT_METHOD_NAME,
-                                       std::vector<code::Expr*> { new code::StringLiteral(qr->str()) }));
+            BOOL_TYPE : mapIRTypeToCodeType(qr->getVar()->getTyp())) }), false, true);
+    code::VarDecl::createVarDecl(coreNs, answervarname,
+      code::Type(DYNAMICTABLE_CLASS_NAME,
+         std::vector<code::Type>{histType, code::Type(std::to_string(qr->getArgs().size()))}));
+
+    std::vector<code::Expr*> args;
+    args.push_back(transExpr(qr->getVar()));
+    args.push_back(new code::IntegerLiteral(1));
+    code::Stmt* evalSt = createPtrMethodCall(
+        lhs2, HISTOGRAM_ADD_METHOD_NAME, args);
+    code::Stmt* printSt = new code::CompoundStmt();
+    std::vector<code::Expr*> printExprVec = std::vector<code::Expr*>( {
+        new code::Identifier(buffername)});
+    std::string typeArgs = qr->getArg(0)->getTyp()->toString() + "[%d]";
+    for (int k = 1; k <= (int)qr->getArgs().size() - 1; k++) {
+        typeArgs += ", " + qr->getArg(k)->getTyp()->toString() + "[%d]";
+    }
+
+    // TODO: Find a new way to acquire just the query variable
+    // HACK FOR NOW: Acquire the name of the variable through string parsing
+    std::string variableName = qr->str().substr(0, qr->str().find("("));
+    printExprVec.push_back(new code::StringLiteral(variableName + "(" + typeArgs + ")\\n"));
+    for (int i = 0; i < argName.size(); i++) {
+        printExprVec.push_back(new code::Identifier(argName[i]));
+    }
+    ((code::CompoundStmt *) printSt)->addStmt(
+        new code::CallExpr(
+            new code::Identifier("sprintf"), printExprVec));
+    ((code::CompoundStmt *) printSt)->addStmt(
+        createPtrMethodCall(
+            lhs3, HISTOGRAM_PRINT_METHOD_NAME,
+            std::vector<code::Expr*> { new code::Identifier(buffername) }));
+    if (qr->getCond() != nullptr) {
+        evalSt = new code::IfStmt(transExpr(qr->getCond()), evalSt);
+        printSt = new code::IfStmt(transExpr(qr->getCond()), printSt);
+    }
+    // generate eval and print for-loops
+    for (int k = (int)qr->getArgs().size() - 1; k >= 0; --k) {
+      auto ty = dynamic_cast<const ir::NameTy*>(qr->getArg(k)->getTyp());
+      assert(ty != NULL);
+      code::Expr* loop_n = NULL; // loop_n duplicated
+      code::Expr* loop_n2 = NULL;
+      if (ty->getRefer()->getNumberStmtSize() == 0) {
+        loop_n = new code::IntegerLiteral(ty->getRefer()->getPreLen());
```

Comment:
Similar to the case with lhs, loop_n is duplicated since it is subject to a double free.